openforcefield/topology/molecule.py

```diff
@@ -3176,6 +3306,222 @@ def from_openeye(oemol, allow_undefined_stereo=False):
         return toolkit.from_openeye(
             oemol, allow_undefined_stereo=allow_undefined_stereo)
 
+    def to_qcshema(self, multiplicity=1, conformer=0):
+        """
+        Generate the qschema input format used to submit jobs to archive
+        or run qcengine calculations locally
+        spec can be found here <https://molssi-qc-schema.readthedocs.io/en/latest/index.html>
+
+        :return: Dict[symbols: list[str], geometry: np.array(float).bohr, connectivity: list[tuple(int, int, int)],
+          molecular_charge: int, molecular_multiplicity: int]
+
+        Example
+        -------
+
+        Create and validate a qcelemental input
+
+        >>> import qcelemental as qcel
+        >>> mol = Molecule.from_smiles('CC')
+        >>> mol.generate_conformers(n_conformers=1)
+        >>> qcel_mol = qcel.models.Molecule.from_data(mol.to_qcshema())
+        """
+
+        try:
+            geometry = self.conformers[conformer].in_units_of(unit.bohr)
+        except TypeError:
+            raise InvalidConformerError('The molecule must have a conformation to produce a valid qcshema.')
+
+        # Gather the required qschema data
+        charge = sum([atom.formal_charge for atom in self.atoms])
+        connectivity = [(bond.atom1_index, bond.atom2_index, bond.bond_order) for bond in self.bonds]
+        symbols = [Element.getByAtomicNumber(atom.atomic_number).symbol for atom in self.atoms]
+
+        return {'symbols': symbols, 'geometry': geometry, 'connectivity': connectivity,
+                'molecular_charge': charge, 'molecular_multiplicity': multiplicity}
+
+    @classmethod
+    def from_mapped_smiles(cls, mapped_smiles):
+        """
+        Create an openforcefield.topology.molecule.Molecule from a mapped SMILES made with cmiles.
+        The molecule will be in the order of the indexing in the mapped smiles string.
+        :param mapped_smiles: a cmiles style mapped smiles string with explicit hydrogens.
+        :return: openforcefield.topology.molecule.Molecule
+        """
+
+        # extract the atomic mapping
+        mapping = []
+        for atom in mapped_smiles.split(':'):
+            try:
+                mapping.append(int(atom[:2]))
+            except ValueError:
+                try:
+                    mapping.append(int(atom[:1]))
+                except ValueError:
+                    pass
```

Comment:
OpenEye and RDKit should be able to read mapped SMILES and annotate the resulting Molecule objects with "map indices". I think it'd be better to let commercial-grade toolkit handle the mapping, rather than parsing the string ourselves.

Reply:
Good point I have now followed @ChayaSt method in cmiles to extract the atom maps if they are present.